### PR TITLE
Extracted time_series_regression_head

### PR DESCRIPTION
**Pull request by terrytangyuan**
Extracted the head for time series estimators. This is the baby step to make sure the implementation is modeled after the ones in `tf.estimator.canned`. 
cc: @allenlavoie @martinwicke

### CODE REVIEW

File: tensorflow/contrib/timeseries/python/timeseries/head.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import state_ops
+from tensorflow.python.ops import variable_scope
+from tensorflow.python.util import nest
+
+
+def time_series_regression_head(
+        model, state_manager, optimizer, input_statistics_generator=None):
+  """Creates a `_Head` for regression using the mean squared loss.
```
Comment (allenlavoie) [inline tensorflow/contrib/timeseries/python/timeseries/head.py]: Maybe just delete " using the mean squared loss". This is determined in the model, and many models use log likelihood rather than squared loss.

File: tensorflow/contrib/timeseries/python/timeseries/head.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+  Args:
+    model: The object (inheriting from Model) to create a function for.
+    state_manager: A state manager to wrap the model with (or
+        PassthroughStateManager if no state needs to be managed).
+    optimizer: An instance of `tf.train.Optimizer` to use for training.
+    input_statistics_generator: An InputStatisticsFromMiniBatch object from
+        math_utils.py, used for collecting statistics about input data during
+        training.
+  Returns:
+    The model function, suitable for passing to a tf.estimator.Estimator.
```
Comment (allenlavoie) [inline tensorflow/contrib/timeseries/python/timeseries/head.py]: Move the "Args:" to \_\_init__. The docstring and "Returns:" don't make much sense either place; it would be nice to write a one-line docstring for each.

File: tensorflow/contrib/timeseries/python/timeseries/head.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    weight_column: A string or a `_NumericColumn` created by
+      `tf.feature_column.numeric_column` defining feature column representing
+      weights. It is used to down weight or boost examples during training. It
+      will be multiplied by the loss of the example.
+    label_dimension: Number of regression labels per example. This is the size
+      of the last dimension of the labels `Tensor` (typically, this has shape
+      `[batch_size, label_dimension]`).
+
+  Returns:
+    An instance of `_Head` for linear regression.
```
Comment (allenlavoie) [inline tensorflow/contrib/timeseries/python/timeseries/head.py]: In fact it sounds like the rest of this function's documentation needs re-writing for time series rather than linear regression. You can just use the "Args:" from _TimeSeriesRegressionHead.

File: tensorflow/contrib/timeseries/python/timeseries/estimators.py [hunk 4]
```diff
@@ -22,6 +22,7 @@
 from tensorflow.contrib.timeseries.python.timeseries import feature_keys
 from tensorflow.contrib.timeseries.python.timeseries import math_utils
 from tensorflow.contrib.timeseries.python.timeseries import model_utils
```
Comment (allenlavoie) [inline tensorflow/contrib/timeseries/python/timeseries/estimators.py]: The model_utils import should now be unused here, right? There is a model_utils.state_to_dictionary call, but that got moved and looks like it needs updating. You could even move the serving input receiver stuff to the head lib if you wanted.

File: tensorflow/contrib/timeseries/python/timeseries/head.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+    for _, key, _ in numbered_state:
+      del features[key]
+    numbered_state.sort(key=lambda number, *_: number)
+    features[feature_keys.State.STATE_TUPLE] = nest.pack_sequence_as(
+      structure=self.model.get_start_state(),
+      flat_sequence=[tensor for _, _, tensor in numbered_state])
+    return features, True
+
+  def create_estimator_spec(
+          self, features, mode, logits=None, labels=None, train_op_fn=None):
```
Comment (allenlavoie) [inline tensorflow/contrib/timeseries/python/timeseries/head.py]: One-line docstring like "Performs basic error checking and returns an EstimatorSpec."

File: tensorflow/contrib/timeseries/python/timeseries/estimators.py [hunk 6]
```diff
@@ -22,6 +22,7 @@
...
+from tensorflow.contrib.timeseries.python.timeseries import ts_head_lib
```
Comment (allenlavoie) [inline tensorflow/contrib/timeseries/python/timeseries/estimators.py]: Should this be "from tensorflow.contrib.timeseries.python.timeseries import head as ts_head_lib"?

**Review by allenlavoie — COMMENTED** [review]
Neat! Some mostly documentation-related comments.

Also looks like there's a lint error about _TimeSeriesRegressionHead being abstract: https://ci.tensorflow.org/job/tensorflow-pull-requests-sanity/7978/console

And finally you'll need to update the BUILD file (add a py_library rule for head.py, add a py_test rule for the new test file).

**Comment by terrytangyuan** [comment]
@allenlavoie Thanks! I think I've addressed your comments and hopefully the build has been fixed.

**Comment by sb2nov** [comment]
Jenkins, test this please.

**Comment by allenlavoie** [comment]
I think the BUILD file needs to be formatted with [buildifier](https://github.com/bazelbuild/buildtools#buildifier) (sorry that CI test error is awful; there's already a py_test import). Then hopefully the tests will pass and I will take another look.

**Comment by terrytangyuan** [comment]
Thanks! That's useful tip to know (so that I don't have to order those deps myself)! I think the sanity check is passing now. Let's see if other tests are passing as well.

**Comment by allenlavoie** [comment]
Looks like you may need a dependency on the head py_library from the estimators py_library. Not that you have to, but you can also run these tests locally with "bazel run //tensorflow/contrib/timeseries/python/timeseries:head_test" (or go to the directory and just specify :head_test). If that's passing you can run all of the timeseries tests locally with "bazel test //tensorflow/contrib/timeseries/...".

**Comment by terrytangyuan** [comment]
Just added that! I will run bazel test if this still fails (away from my machine currently) and make sure to do it next time before PR as well. Thank you for your patience (haven't done any major dev here for a while)!

**Comment by terrytangyuan** [comment]
@allenlavoie I think I fixed everything now (at least locally). I left a couple of TODOs in the code. I'll try work on those in separate PRs when I have more time!

File: tensorflow/tools/ci_build/ci_sanity.sh [hunk 7]
```diff
@@ -95,6 +95,7 @@ do_pylint() {
 "^tensorflow/python/platform/default/_googletest\.py.*\[E0102.*function\salready\sdefined "\
 "^tensorflow/python/feature_column/feature_column_test\.py.*\[E0110.*abstract-class-instantiated "\
 "^tensorflow/contrib/layers/python/layers/feature_column\.py.*\[E0110.*abstract-class-instantiated "\
+"^tensorflow/contrib/timeseries/python/timeseries/head\.py.*\[E0110.*abstract-class-instantiated "\
```
[the remark this answers is not in the document]
Comment (terrytangyuan) [reply tensorflow/tools/ci_build/ci_sanity.sh]: Good catch! I'll remove it.

[the remark this answers is not in the document]
Comment (terrytangyuan) [reply tensorflow/tools/ci_build/ci_sanity.sh]: I am seeing something new - `Ubuntu CC — Internal CI build failed`. I doubt that it's caused by my last commit for removing ts_head from pylint whitelist though.

File: tensorflow/contrib/timeseries/python/timeseries/head.py [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+      `[batch_size, label_dimension]`).
+
+  Returns:
+    An instance of `_Head` for time series regression.
+  """
+  return _TimeSeriesRegressionHead(
+    model, state_manager, optimizer, input_statistics_generator)
+
+
+class _TimeSeriesRegressionHead(head_lib._Head):  # pylint:disable=protected-access
```
[the remark this answers is not in the document]
Comment (allenlavoie) [reply tensorflow/contrib/timeseries/python/timeseries/head.py]: Sounds like this is fine (_Head may change, but we can deal with that).

**Review by allenlavoie — APPROVED** [review]
Nice work!

File: tensorflow/contrib/timeseries/python/timeseries/head.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (terrytangyuan) [reply tensorflow/contrib/timeseries/python/timeseries/head.py]: Sounds good! BTW, should I pay attention to this `Ubuntu CC — Internal CI build failed`?

**Comment by allenlavoie** [comment]
Jenkins, test this please.

Should be fine; was probably just a one-off tool failure.

**Comment by sb2nov** [comment]
Jenkins, test this please. Probably a flake, retesting

**Comment by sb2nov** [comment]
@gunan what is Ubuntu CC, restarting didn't seem to kick that off again.

**Comment by gunan** [comment]
That is a replacement we are building for jenkins.
It will be triggered using labels.

**Comment by terrytangyuan** [comment]
Triggering run via labels is super neat! Thanks it's finally passing.
